src/Compilers/CSharp/Test/Semantic/Semantics/FunctionPointerTests.cs

```diff
@@ -726,5 +727,91 @@ void M(delegate*<delegate*<object, void>, void> param1, delegate*<delegate*<obje
+        delegate*<string, void> p = null;
+        M1(p);
+    }
+}");
+
+            comp.VerifyDiagnostics();
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var invocation = tree.GetRoot().DescendantNodes().OfType<InvocationExpressionSyntax>().Single();
+            var methodSymbol = (IMethodSymbol)model.GetSymbolInfo(invocation).Symbol!;
+
+            Assert.NotSame(methodSymbol, methodSymbol.OriginalDefinition);
+            Assert.Equal(SpecialType.System_String, methodSymbol.TypeArguments[0].SpecialType);
+            var functionPointer = (FunctionPointerTypeSymbol)methodSymbol.Parameters[0].Type;
+            Assert.Equal(SpecialType.System_String, functionPointer.Signature.Parameters[0].Type.SpecialType);
+        }
+
+        [Fact]
+        public void FunctionPointerReturnTypeInference()
+        {
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class C
+{
+    public void M1<T>(delegate*<T> param) {}
+    public void M2()
+    {
+        delegate*<string> p = null;
+        M1(p);
+    }
+}");
+
+            comp.VerifyDiagnostics();
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var invocation = tree.GetRoot().DescendantNodes().OfType<InvocationExpressionSyntax>().Single();
+            var methodSymbol = (IMethodSymbol)model.GetSymbolInfo(invocation).Symbol!;
+
+            Assert.NotSame(methodSymbol, methodSymbol.OriginalDefinition);
+            Assert.Equal(SpecialType.System_String, methodSymbol.TypeArguments[0].SpecialType);
+            var functionPointer = (FunctionPointerTypeSymbol)methodSymbol.Parameters[0].Type;
+            Assert.Equal(SpecialType.System_String, functionPointer.Signature.ReturnType.SpecialType);
+        }
+
+        [Fact]
+        public void FunctionPointerGenericSubstitutionInference()
+        {
+
+            var comp = CreateCompilationWithFunctionPointers(@"
+unsafe class C
+{
+    public void M1<T>(delegate*<T, T> param) {}
+    public void M2<T>()
+    {
+        delegate*<T, T> p = null;
+        M1(p);
+    }
+}");
+
+            comp.VerifyDiagnostics();
+
+            var tree = comp.SyntaxTrees[0];
+            var model = comp.GetSemanticModel(tree);
+
+            var invocation = tree.GetRoot().DescendantNodes().OfType<InvocationExpressionSyntax>().Single();
+            var methodSymbol = (IMethodSymbol)model.GetSymbolInfo(invocation).Symbol!;
+
+            Assert.NotSame(methodSymbol, methodSymbol.OriginalDefinition);
+            Assert.Equal(TypeKind.TypeParameter, methodSymbol.TypeArguments[0].TypeKind);
+            var functionPointer = (FunctionPointerTypeSymbol)methodSymbol.Parameters[0].Type;
+            Assert.Equal(TypeKind.TypeParameter, functionPointer.Signature.ReturnType.TypeKind);
+            Assert.Equal(TypeKind.TypeParameter, functionPointer.Signature.Parameters[0].Type.TypeKind);
+        }
     }
```

Comment:
>} [](start = 4, length = 1)

I think we should also test a scenarios when we infer a function pointer for a type parameter. Including scenarios when there are more than one candidate and the types are:
- identical
- convertible
- not convertible
- with just calling convention differences
- etc.

Reply:
> I think we should also test a scenarios when we infer a function pointer for a type parameter. 

Function pointers cannot be type parameters.

---
In reply to: [406422979](https://github.com/dotnet/roslyn/pull/43041#discussion_r406422979) [](ancestors = 406422979)

Reply:
> Function pointers cannot be type parameters.

But we are going to try to infer something. Let's see what is going to happen and what do we infer.
Also, consider adding tests for inference for array initializers and for conditional expressions, it might just work.

---
In reply to: [406455969](https://github.com/dotnet/roslyn/pull/43041#discussion_r406455969) [](ancestors = 406455969,406422979)

Reply:
Also, we might avoid an error if we infer directly form an array of function pointers rather than from a function pointer. Might want to test scenarios like that as well.

---
In reply to: [406467051](https://github.com/dotnet/roslyn/pull/43041#discussion_r406467051) [](ancestors = 406467051,406455969,406422979)

Reply:
It doesn't looks like there are test covering:
- not convertible (for arrays)
- with just calling convention differences (for arrays and pointers)

---
In reply to: [406468735](https://github.com/dotnet/roslyn/pull/43041#discussion_r406468735) [](ancestors = 406468735,406467051,406455969,406422979)

Reply:
Just calling convention differences are covered by `FunctionPointerAsTypeArgument` for pointer types.

---
In reply to: [406881794](https://github.com/dotnet/roslyn/pull/43041#discussion_r406881794) [](ancestors = 406881794,406468735,406467051,406455969,406422979)

Reply:
> Just calling convention differences are covered by FunctionPointerAsTypeArgument for pointer types.

Please add coverage for arrays too

---
In reply to: [406910516](https://github.com/dotnet/roslyn/pull/43041#discussion_r406910516) [](ancestors = 406910516,406881794,406468735,406467051,406455969,406422979)

Reply:
This is marked as resolved, but I still don't see the tests for method type inference
- not convertible (for arrays)
- with just calling convention differences (for arrays)

---
In reply to: [406933066](https://github.com/dotnet/roslyn/pull/43041#discussion_r406933066) [](ancestors = 406933066,406910516,406881794,406468735,406467051,406455969,406422979)